packages/firestore/test/integration/api/server_timestamp.test.ts

```diff
@@ -158,7 +200,177 @@ apiDescribe('Server Timestamps', persistence => {
             return Promise.resolve();
           })
         )
-        .then(() => waitForRemoteEvent());
+        .then(waitForRemoteEvent)
+        .then(snapshot => verifyTimestampsAreResolved(snapshot));
+    });
+  });
+
+  it('can return estimated value', () => {
+    return withTestSetup(() => {
+      return writeInitialData()
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot => verifyTimestampsAreEstimates(snapshot))
+        .then(waitForRemoteEvent)
+        .then(snapshot => verifyTimestampsAreResolved(snapshot));
+    });
+  });
+
+  it('can return previous value', () => {
+    let previousSnapshot: firestore.DocumentSnapshot;
+
+    return withTestSetup(() => {
+      return writeInitialData()
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot => verifyTimestampsUsePreviousValue(snapshot, null))
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          previousSnapshot = snapshot;
+          verifyTimestampsAreResolved(snapshot);
+        })
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot =>
+          verifyTimestampsUsePreviousValue(snapshot, previousSnapshot)
+        )
+        .then(waitForRemoteEvent)
+        .then(snapshot => verifyTimestampsAreResolved(snapshot));
+    });
+  });
+
+  it('can return previous value of different type', () => {
+    return withTestSetup(() => {
+      return writeInitialData()
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot => verifyTimestampsUsePreviousValue(snapshot, null))
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          verifyTimestampsAreResolved(snapshot);
+        })
```

Comment:
Lines 245-251 are repeated from the previous test, right?  Why?  It's really hard to read tests that are long and repetitive. I can't figure out what's setup vs behavior being tested (comments can be used as a crutch, but you didn't do that either :-)).

My preferred way to do this test (in the spirit of DRY) would probably be to change writeInitialData() to accept an optional initial value for the when fields.

Then you can just do writeInitialData(/*whenValue=*/42).then(() => docRef.update(updateData)) and then wait for the events / do your validation.

Reply:
I cleaned up the tests quite a bit and removed the parts that are exercised in other tests. I think this reduces the scope and duplication enough to not introduce further helpers.